packages/itwinui-react/src/core/SearchBox/SearchBox.tsx

```diff
@@ -0,0 +1,398 @@
+      onClose?: () => void;
+      /**
+       * Callback for expanding searchbox
+       */
+      onOpen?: () => void;
+    }
+  | undefined
+>(undefined);
+
+// ----------------------------------------------------------------------------
+
+type SearchBoxOwnProps = {
+  /**
+   * Whether the searchbox is expandable.
+   * @default false
+   */
+  expandable?: boolean;
+  /**
+   * Searchbox state toggle.
+   */
+  isExpanded?: boolean;
+  /**
+   * Function that is called on toggling (expand, collapse) searchbox.
+   */
+  onToggle?: (isExpanding: boolean) => void;
+  /**
+   * Collapsed searchbox state.
+   * @default SearchBox.OpenButton
+   */
+  collapsedState?: React.ReactNode;
+  /**
+   * Input props when using default searchbox.
+   */
+  inputProps?: React.ComponentPropsWithoutRef<'input'>;
+  /**
+   * Open button props when using default searchbox.
+   */
+  openButtonProps?: React.ComponentPropsWithoutRef<'button'>;
+  /**
+   * Close button props when using default searchbox.
+   */
+  closeButtonProps?: React.ComponentPropsWithoutRef<'button'>;
+  /**
+   * Modify size of the searchbox and it's subcomponents.
+   */
+  size?: 'small' | 'large';
+};
+
+const SearchBoxComponent = React.forwardRef((props, ref) => {
+  useTheme();
+
+  const {
+    size,
+    expandable = false,
+    disabled = false,
+    onToggle,
+    collapsedState,
+    isExpanded,
+    children,
+    inputProps,
+    openButtonProps,
+    closeButtonProps,
+    className,
+    ...rest
+  } = props;
+
+  const [inputId, setInputId] = React.useState(useId());
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const openButtonRef = React.useRef<HTMLButtonElement>(null);
+
+  const [localExpanded, setLocalExpanded] = React.useState(isExpanded);
+
+  const onClose = () => {
+    setLocalExpanded(false);
+    onToggle?.(false);
+    openButtonRef.current && openButtonRef.current.focus();
+  };
+
+  const onOpen = () => {
+    setLocalExpanded(true);
+    onToggle?.(true);
+    inputRef.current && inputRef.current.focus();
+  };
+
+  return (
+    <SearchBoxContext.Provider
+      value={{
+        size,
+        disabled,
+        onOpen,
+        onClose,
+        inputRef,
+        inputId,
+        setInputId,
+        openButtonRef,
+      }}
+    >
+      <InputFlexContainer
+        ref={ref}
+        className={cx({ 'iui-expandable-searchbox': expandable }, className)}
+        data-iui-size={size}
+        disabled={disabled}
+        role='searchbox'
+        data-iui-expanded={localExpanded}
+        {...rest}
+      >
+        {!expandable &&
+          (children ?? (
+            <>
+              <SearchBoxInput {...inputProps} />
+              <SearchBoxIcon />
+            </>
+          ))}
+        {expandable && (
+          <>
+            <div className='iui-searchbox-collapsed'>
+              {collapsedState ?? <SearchBoxOpenButton {...openButtonProps} />}
+            </div>
+            <div className='iui-searchbox-expanded'>
```

Comment:
there is no need for `iui-searchbox-collapsed` and `iui-searchbox-expanded` classes. It should be enough to remove/add the actual elements from the DOM like this:

```
!expanded ? collapsedState : expandedState
```

Reply:
Thanks! Updated code to use this structure.
Added isExpanded and localExpanded checks for manual control.

Reply:
you can create a new variable and use that as the source of truth everywhere
![](https://user-images.githubusercontent.com/9084735/235960257-29aa2402-538c-4c73-82b5-9056513a0e7d.png)

Reply:
Using this magic var.

Reply:
> * Did we ever decide on whether default basic state should have search icon at the end or start?

UX team was asked this question. If I remember correctly, no direct answer provided. There are more cases of search icon at the end of the searchbox in products that our company creates, based on research (articles and examples in wide world of internet) - it's better to have search icon at the start. @FlyersPh9 Do you have any thoughts?

> * For expandable search, what is the expected way to set width?

Expandable search takes width of the container or set maxWidth on itself?
With grid:

https://user-images.githubusercontent.com/83585998/236450725-42b62e4b-0870-4d91-9f37-faca45f8e634.mov

With PageLayout:

https://user-images.githubusercontent.com/83585998/236450772-7c9dc09d-d92b-4ebb-b1aa-05eb67204d37.mov

With maxWidth:

https://user-images.githubusercontent.com/83585998/236450796-d19b4d35-9215-4b4e-9d98-6915d7a57d77.mov


> * Should we have any guidance on where/how expandable search should be used? Because if there are items next to the searchbox, then it is going to cause a lot of layout shift and it is also very expensive for the browser to compute.

Yes, we should. I found in old patterns site that expandable searchbox goes OVER content next to it. Will look into it more and add examples of usage in docs website.

> * Do we have any precedent on the animation? Right now the animation feels too slow and linear. Could change the duration and the easing function. (question for @FlyersPh9 i guess)

Answered in FlyersPh9 comment.

> * Are we going to expose InputFlexContainer publicly in v2? Or does it need more work and we should save it for after initial v3 release?

It should be good to be exposed, if we want to expose it.

Reply:
This was worked on in pair program session.